src/runtime/arch_wasm.go

```diff
@@ -30,3 +33,32 @@ func align(ptr uintptr) uintptr {
 }
 
 func getCurrentStackPointer() uintptr
+
+// growHeap tries to grow the heap size. It returns true if it succeeds, false
+// otherwise.
+func growHeap() bool {
+	// Grow memory by the available size, which means the heap size is doubled.
+	memorySize := wasm_memory_size(0)
+	result := wasm_memory_grow(0, memorySize)
+	if result == -1 {
+		// Grow failed.
+		return false
+	}
+
+	// Save some old variables we need later.
+	oldMetadataStart := metadataStart
+	oldMetadataSize := heapEnd - uintptr(metadataStart)
+
+	// Increase the heap. After setting the new heapEnd, calculateHeapAddresses
+	// will update metadataStart and the memcpy will copy the metadata to the
+	// new location.
+	// The new metadata will be bigger than the old metadata, but a simple
+	// memcpy is fine as it only copies the old metadata and the new memory will
+	// have been zero initialized.
+	heapEnd = uintptr(wasm_memory_size(0) * wasmPageSize)
+	calculateHeapAddresses()
+	memcpy(unsafe.Pointer(metadataStart), unsafe.Pointer(oldMetadataStart), oldMetadataSize)
```

Comment:
This section should probably live inside of `gc_conservative.go`, as this references variables from a separate build tag. We could maybe have something like `resizeHeap` (I am not doing great with names right now) which would sit inside of the GC implementations. If the GC did not support this, it could avoid calling `growHeap` and implement `resizeHeap` as a no-op. Ideally, `leaking`/similar should still work.

Reply:
That seems like a very reasonable suggestion to me. :+1:

Reply:
Yes, that makes sense. I have updated the PR with this change (with `setHeapEnd`) that is also supported with `-gc=none` and `-gc=leaking` (it doesn't make sense with `-gc=extalloc`).